### PR TITLE
support multiple header values

### PR DESCRIPTION
**Pull request by ME-ON1**
Signed-off-by: Tarun Sharma <starun.1998@gmail.com>

Commit Message: support multiple header values
Additional Description: #13454
Risk Level: N/A
Testing: locally
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

File: source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc
Comment (wbpcode) [inline source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:122]: We can remove this TODO now.

File: source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc
Comment (wbpcode) [inline source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:147]: ditto.

File: source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc
Comment (wbpcode) [inline source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:162]: ditto.

File: source/extensions/filters/network/thrift_proxy/router/router_ratelimit_impl.cc
Comment (wbpcode) [inline source/extensions/filters/network/thrift_proxy/router/router_ratelimit_impl.cc:48]: ditto.

File: source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc
Comment (wbpcode) [inline source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:145]: I think this if judgment is now meaningless.

File: source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc
Comment (wbpcode) [inline source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:120]: I think this if judgment is now meaningless.

**Review by wbpcode — CHANGES_REQUESTED** [review]
Thanks for your contribution. 😄 And can you add some more test to cover the case that the single header with multiple values or single value.

File: source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc
Comment (wbpcode) [inline source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:125]: Assuming that logged_headers is just a normal map, it still seems that only the first header will be successfully inserted. I think this may not achieve our purpose.

File: source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc
Comment (wbpcode) [inline source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:150]: ditto.

File: source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc
Comment (wbpcode) [inline source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:165]: ditto.

File: source/extensions/filters/network/thrift_proxy/router/router_ratelimit_impl.cc
Comment (wbpcode) [inline source/extensions/filters/network/thrift_proxy/router/router_ratelimit_impl.cc:52]: I'm not sure if this will affect the users of the current rate-limit function.

**Review by wbpcode — COMMENTED** [review]
Some more comments. 😃

File: source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc
Comment (ME-ON1) [reply source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:125]: if this is map i think then we just have to iterate map values and insert it ?!

File: source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc
Comment (ME-ON1) [reply source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:122]: sure!

File: source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc
Comment (wbpcode) [reply source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:125]: What I mean is that inserting multiple header values into the map with the same key does not get the result we expect: save all header values. 🤔

Comment (ME-ON1) [reply source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:125]: ohh yeah got it. what should be the key values then ?!

Comment (wbpcode) [reply source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:125]: I also don't have any good ideas, this problem is not easy.  🤔

Comment (ME-ON1) [reply source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:125]: ohh!! 
 @lizan  do you have idea for this mapped values logging ?!

Comment (wbpcode) [reply source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:125]: @ME-ON1 May be `HeaderUtility::getAllOfHeaderAsString` is a possible solution.

Comment (ME-ON1) [reply source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:125]: for inserting the values in logged_headers map ?

Comment (wbpcode) [reply source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:125]: Yep. But But it is best to develop after the community accepts such a program.

Comment (ME-ON1) [reply source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:125]: whom should i ask for GO !?

Comment (lizan) [reply source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:125]: Please add a test so it is much easier to see what's expected here for behaviors. `getAllOfHeaderAsString` could work if we decided in that way and documented correctly.

File: source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc
Comment (wbpcode) [reply source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:148]: Here is previous comments.

File: source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc
Comment (wbpcode) [inline source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:150]: It seems that it is unnecessary to convert `header` to `Http::LowerCaseString` with `Http::LowerCaseString(header)`. The type of header was originally  `Http::LowerCaseString`.

File: source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc
Comment (wbpcode) [inline source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:151]: May be `value_or()`?

File: source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc
Comment (wbpcode) [inline source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:166]: ditto.

File: source/extensions/filters/network/thrift_proxy/router/router_ratelimit_impl.cc
Comment (wbpcode) [reply source/extensions/filters/network/thrift_proxy/router/router_ratelimit_impl.cc:52]: Here is previous comment.

**Review by wbpcode — CHANGES_REQUESTED** [review]
Thanks for you contributions. And some comments are added and test is need. Thanks.

File: source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc
Comment (wbpcode) [inline source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:125]: may be just:

```
        const auto all_values = Http::HeaderUtility::getAllOfHeaderAsString(
            request_headers, header);
	if(all_values.result().has_value()) {
        	logged_headers->insert({header.get(), std::string(all_values.result().value())});
	}
 ```

**Comment by wbpcode** [comment]
Thanks, some minor comments are added. And we can fix the format error by the `ci/do_ci.sh fix_format`.

/wait

**Comment by ME-ON1** [comment]
@wbpcode hey i did run this cmd before committing and also this `ci/run_envoy_docker.sh 'ci/do_ci.sh format_pre`  and its running without error locally !! ?

**Comment by wbpcode** [comment]
> @wbpcode hey i did run this cmd before committing and also this `ci/run_envoy_docker.sh 'ci/do_ci.sh format_pre` and its running without error locally !! ?

Then do you push the latest commits? You can check the CI detail log to get more info about the format error:

```
##[error]: mixed tabs and spaces: source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc
##[error]: mixed tabs and spaces: test/extensions/access_loggers/grpc/http_grpc_access_log_impl_test.cc
```

**Comment by ME-ON1** [comment]
made another commit with all these and ran all these cmds and no errors locally.

File: source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc
Comment (wbpcode) [inline source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:145]: Because we has ensured tht the `result()` has a value so there is unnecessary to use `value_or`.

**Review by wbpcode — COMMENTED** [review]
Thanks. It looks good overall and only one minor comment.

File: source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc
Comment (ME-ON1) [reply source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:145]: sure should use `value` now

Comment (ME-ON1) [reply source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc]: done

File: source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc
Comment (wbpcode) [inline source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:154]: It seems that this is not needed anymore.

File: source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc
Comment (wbpcode) [inline source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:142]: ditto.

**Comment by wbpcode** [comment]
Hi, @ME-ON1 , can you check the lastest comments. 🤣

**Comment by alyssawilk** [comment]
@wbpcode as a reminder if you /wait or tag as waiting (when PRs are waiting for author work) this won't nag you or us in tooling :-)
/wait

File: source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc
Comment (ME-ON1) [reply source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc]: yes. rm-ed

File: source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc
Comment (ME-ON1) [reply source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc]: done

**Review by wbpcode — DISMISSED** [review]
/lgtm

cc @envoyproxy/envoy-maintainers for second pass.

**Comment by mattklein123** [comment]
@lizan can you take a pass on this?

**Review by lizan — CHANGES_REQUESTED** [review]
Code LGTM, can you add a release note of this behavior change?

**Comment by snowp** [comment]
Seems like this one is waiting for a release note to be added?

/wait

**Comment by wbpcode** [comment]
You may need add one line note at the `docs/root/version_history/current.rst#Minor Behavior Changes` to record your changes.

For example:
```
* access_log: log all header values in the grpc access log.
```

Note: Pay attention to the order of the notes.

**Comment by ME-ON1** [comment]
@wbpcode @lizan PTAL

**Review by wbpcode — APPROVED** [review]
Thanks, LGTM. cc @lizan

**Comment by wbpcode** [comment]
Friendly ping @lizan PTAL for this PR, thanks. 😄

**Review by lizan — APPROVED** [review]

**Pull request merged**
